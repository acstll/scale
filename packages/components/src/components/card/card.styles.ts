import { JssStyle } from 'jss';

const styles: JssStyle = {
  card: {
    boxSizing: 'border-box',
    overflow: 'hidden',
    color: '#333',
    background: '#fff',
    border: '1px solid rgba(0, 0, 0, 0.1)',
    borderRadius: '4px',
    width: '400px',
    maxWidth: '550px',
    boxShadow: '0 2px 4px 0 rgba(0, 0, 0, 0.12)',
  },
  card__body: {
    padding: '1rem',
  },
  card__header: {
    padding: '0.5rem 1rem',
    borderBottom: '1px solid, #dfdfdf',
    background: '#eee',
    fontSize: '0.8rem',
    '&::slotted(*)': {
      margin: 0,
    },
  },
  card__footer: {
    padding: '0.5rem 1rem',
    borderTop: '1px solid #dfdfdf',
    background: '#eee',
    fontSize: '0.8rem',
    '&::slotted(*)': {
      margin: 0,
    },
  },
  'card__img-top': {},
};

export default styles;
