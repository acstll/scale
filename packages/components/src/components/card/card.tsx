import { Component, Prop, h, Element, Host } from '@stencil/core';
import { StyleSheet } from 'jss';
import { CssClassMap } from '../../utils/utils';
import classNames from 'classnames';
import Base from '../../utils/base-interface';
import { CssInJs } from '../../utils/decorators/css-in-js';
import defaultStyles from './card.styles';

@Component({
  tag: 't-card',
  shadow: true,
})
export class Card implements Base {
  @Element() hostElement: HTMLElement;
  /** (optional) Card class */
  @Prop() customClass?: string = '';
  /** (optional) Card size */
  @Prop() size?: string = '';
  /** (optional) Card variant */
  @Prop() variant?: string = '';
  /** (optional) Disabled card */
  @Prop() disabled?: boolean = false;
  /** (optional) Deselected card */
  @Prop() deselected?: boolean = false;
  /** (optional) Card image at the top */
  @Prop() imageTop?: string;
  /** (optional) Card image alternative at the top */
  @Prop() imageTopAlt?: string = '';

  /** (optional) Injected jss styles */
  @Prop() styles?: StyleSheet;
  /** decorator Jss stylesheet */
  @CssInJs('Card', defaultStyles) stylesheet: StyleSheet;
  hasSlotHeader: boolean;
  hasSlotFooter: boolean;

  componentWillLoad() {
    this.hasSlotHeader = !!this.hostElement.querySelector('[slot="header"]');
    this.hasSlotFooter = !!this.hostElement.querySelector('[slot="footer"]');
  }

  render() {
    const { classes } = this.stylesheet;
    return (
      <Host>
        <style>{this.stylesheet.toString()}</style>
        <div class={this.getCssClassMap()}>
          {this.hasSlotHeader && (
            <div class={classes.card__header}>
              <slot name="header" />
            </div>
          )}
          {this.imageTop && (
            <img
              class={classes['card__img-top']}
              src={this.imageTop}
              alt={this.imageTopAlt}
            />
          )}
          <div class={classes.card__body}>
            <slot />
          </div>
          {this.hasSlotFooter && (
            <div class={classes.card__footer}>
              <slot name="footer" />
            </div>
          )}
        </div>
      </Host>
    );
  }

  getCssClassMap(): CssClassMap {
    const { classes } = this.stylesheet;
    return classNames(
      classes.card,
      this.customClass && this.customClass,
      this.size && `${classes.card}--size-${this.size}`,
      this.variant && `${classes.card}--variant-${this.variant}`,
      this.disabled && `${classes.card}--disabled`,
      this.deselected && `${classes.card}--deselected`
    );
  }
}
