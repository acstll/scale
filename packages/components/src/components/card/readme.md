# t-card



<!-- Auto Generated Below -->


## Properties

| Property      | Attribute       | Description                                  | Type                                     | Default     |
| ------------- | --------------- | -------------------------------------------- | ---------------------------------------- | ----------- |
| `customClass` | `custom-class`  | (optional) Card class                        | `string`                                 | `''`        |
| `deselected`  | `deselected`    | (optional) Deselected card                   | `boolean`                                | `false`     |
| `disabled`    | `disabled`      | (optional) Disabled card                     | `boolean`                                | `false`     |
| `imageTop`    | `image-top`     | (optional) Card image at the top             | `string`                                 | `undefined` |
| `imageTopAlt` | `image-top-alt` | (optional) Card image alternative at the top | `string`                                 | `''`        |
| `size`        | `size`          | (optional) Card size                         | `string`                                 | `''`        |
| `styles`      | --              | (optional) Injected jss styles               | `StyleSheet<string \| number \| symbol>` | `undefined` |
| `variant`     | `variant`       | (optional) Card variant                      | `string`                                 | `''`        |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
