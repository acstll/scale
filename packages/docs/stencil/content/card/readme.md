# scale-card



<!-- Auto Generated Below -->


## Usage

### Card

# card

<scale-card>
  <h1>Card Title</h1>
  <p>This is card content</p>
</scale-card>



## Properties

| Property      | Attribute      | Description                    | Type                                     | Default     |
| ------------- | -------------- | ------------------------------ | ---------------------------------------- | ----------- |
| `customClass` | `custom-class` | (optional) Card class          | `string`                                 | `''`        |
| `styles`      | --             | (optional) Injected jss styles | `StyleSheet<string \| number \| symbol>` | `undefined` |
| `variant`     | `variant`      | (optional) Card variant        | `string`                                 | `''`        |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
