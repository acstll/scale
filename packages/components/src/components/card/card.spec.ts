import { newSpecPage } from '@stencil/core/testing';
import { Card } from './card';
import jss from 'jss';
import cardStyles from './card.styles';

describe('Card', () => {
  let element;
  beforeEach(async () => {
    element = new Card();
    element.stylesheet = jss.createStyleSheet(cardStyles as any);
  });

  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [Card],
      html: `<t-card>Label</t-card>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should match snapshot with header slot', async () => {
    const page = await newSpecPage({
      components: [Card],
      html: `
			<t-card>
				<h3 slot="header">Header content</h3>
				A title
			</t-card>
			`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should match snapshot with footer slot', async () => {
    const page = await newSpecPage({
      components: [Card],
      html: `
			<t-card>
				<h3 slot="footer">Footer content</h3>
				A title
			</t-card>
			`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should match snapshot with image', async () => {
    const page = await newSpecPage({
      components: [Card],
      html: `<t-card image-top="http://placehold.it/400x300">A title</t-card>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should have a default css class', () => {
    expect(element.getCssClassMap()).toBe(element.stylesheet.classes.card);
  });

  it('should handle size css class', () => {
    element.size = 'small';
    expect(element.getCssClassMap()).toContain(
      `${element.stylesheet.classes.card}--size-small`
    );
  });

  it('should handle variant css class', () => {
    element.variant = 'primary';
    expect(element.getCssClassMap()).toContain(
      `${element.stylesheet.classes.card}--variant-primary`
    );
  });

  it('should handle disabled css class', () => {
    element.disabled = true;
    expect(element.getCssClassMap()).toContain(
      `${element.stylesheet.classes.card}--disabled`
    );
  });

  it('should handle deselected css class', () => {
    element.deselected = true;
    expect(element.getCssClassMap()).toContain(
      `${element.stylesheet.classes.card}--deselected`
    );
  });
});
