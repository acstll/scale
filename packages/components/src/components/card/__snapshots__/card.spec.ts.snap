// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match snapshot 1`] = `
<t-card>
  <mock:shadow-root>
    <style>
      .card-0-0-6 {
  color: #333;
  width: 400px;
  border: 1px solid rgba(0, 0, 0, 0.1);
  overflow: hidden;
  max-width: 550px;
  background: #fff;
  box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.12);
  box-sizing: border-box;
  border-radius: 4px;
}
.card__body-0-0-7 {
  padding: 1rem;
}
.card__header-0-0-8 {
  padding: 0.5rem 1rem;
  font-size: 0.8rem;
  background: #eee;
  border-bottom: 1px solid, #dfdfdf;
}
.card__header-0-0-8::slotted(*) {
  margin: 0;
}
.card__footer-0-0-9 {
  padding: 0.5rem 1rem;
  font-size: 0.8rem;
  background: #eee;
  border-top: 1px solid #dfdfdf;
}
.card__footer-0-0-9::slotted(*) {
  margin: 0;
}
    </style>
    <div class="card-0-0-6">
      <div class="card__body-0-0-7">
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
  Label
</t-card>
`;

exports[`Card should match snapshot with footer slot 1`] = `
<t-card>
  <mock:shadow-root>
    <style>
      .card-0-0-26 {
  color: #333;
  width: 400px;
  border: 1px solid rgba(0, 0, 0, 0.1);
  overflow: hidden;
  max-width: 550px;
  background: #fff;
  box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.12);
  box-sizing: border-box;
  border-radius: 4px;
}
.card__body-0-0-27 {
  padding: 1rem;
}
.card__header-0-0-28 {
  padding: 0.5rem 1rem;
  font-size: 0.8rem;
  background: #eee;
  border-bottom: 1px solid, #dfdfdf;
}
.card__header-0-0-28::slotted(*) {
  margin: 0;
}
.card__footer-0-0-29 {
  padding: 0.5rem 1rem;
  font-size: 0.8rem;
  background: #eee;
  border-top: 1px solid #dfdfdf;
}
.card__footer-0-0-29::slotted(*) {
  margin: 0;
}
    </style>
    <div class="card-0-0-26">
      <div class="card__body-0-0-27">
        <slot></slot>
      </div>
      <div class="card__footer-0-0-29">
        <slot name="footer"></slot>
      </div>
    </div>
  </mock:shadow-root>
  <h3 slot="footer">
    Footer content
  </h3>
  A title
</t-card>
`;

exports[`Card should match snapshot with header slot 1`] = `
<t-card>
  <mock:shadow-root>
    <style>
      .card-0-0-16 {
  color: #333;
  width: 400px;
  border: 1px solid rgba(0, 0, 0, 0.1);
  overflow: hidden;
  max-width: 550px;
  background: #fff;
  box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.12);
  box-sizing: border-box;
  border-radius: 4px;
}
.card__body-0-0-17 {
  padding: 1rem;
}
.card__header-0-0-18 {
  padding: 0.5rem 1rem;
  font-size: 0.8rem;
  background: #eee;
  border-bottom: 1px solid, #dfdfdf;
}
.card__header-0-0-18::slotted(*) {
  margin: 0;
}
.card__footer-0-0-19 {
  padding: 0.5rem 1rem;
  font-size: 0.8rem;
  background: #eee;
  border-top: 1px solid #dfdfdf;
}
.card__footer-0-0-19::slotted(*) {
  margin: 0;
}
    </style>
    <div class="card-0-0-16">
      <div class="card__header-0-0-18">
        <slot name="header"></slot>
      </div>
      <div class="card__body-0-0-17">
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
  <h3 slot="header">
    Header content
  </h3>
  A title
</t-card>
`;

exports[`Card should match snapshot with image 1`] = `
<t-card image-top="http://placehold.it/400x300">
  <mock:shadow-root>
    <style>
      .card-0-0-36 {
  color: #333;
  width: 400px;
  border: 1px solid rgba(0, 0, 0, 0.1);
  overflow: hidden;
  max-width: 550px;
  background: #fff;
  box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.12);
  box-sizing: border-box;
  border-radius: 4px;
}
.card__body-0-0-37 {
  padding: 1rem;
}
.card__header-0-0-38 {
  padding: 0.5rem 1rem;
  font-size: 0.8rem;
  background: #eee;
  border-bottom: 1px solid, #dfdfdf;
}
.card__header-0-0-38::slotted(*) {
  margin: 0;
}
.card__footer-0-0-39 {
  padding: 0.5rem 1rem;
  font-size: 0.8rem;
  background: #eee;
  border-top: 1px solid #dfdfdf;
}
.card__footer-0-0-39::slotted(*) {
  margin: 0;
}
    </style>
    <div class="card-0-0-36">
      <img alt="" class="card__img-top-0-0-40" src="http://placehold.it/400x300">
      <div class="card__body-0-0-37">
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
  A title
</t-card>
`;
